In [81]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os

os.listdir('/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset')
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/female_players_17.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/female_players_20.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_15.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_16.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/Career Mode female player datasets - FIFA 16-22.xlsx
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_17.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_18.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_19.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/female_players_16.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_22.csv
/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/female_pla

### Importing libraries

In [82]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option("display.max_columns",200)

## Read csv file

In [83]:
df = pd.read_csv("/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_22.csv")

/tmp/ipykernel_127/3143148529.py:1: DtypeWarning: Columns (25,108) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("/kaggle/input/datasets/stefanoleone992/fifa-22-complete-player-dataset/players_22.csv")


## Dataset Understanding

* shape
* Head and tail
* describe
* dtype
* null


In [84]:
df.shape

(19239, 110)

In [85]:
df.head()


,sofifa_id,player_url,short_name,long_name,player_positions,overall,potential,value_eur,wage_eur,age,dob,height_cm,weight_kg,club_team_id,club_name,league_name,league_level,club_position,club_jersey_number,club_loaned_from,club_joined,club_contract_valid_until,nationality_id,nationality_name,nation_team_id,nation_position,nation_jersey_number,preferred_foot,weak_foot,skill_moves,international_reputation,work_rate,body_type,real_face,release_clause_eur,player_tags,player_traits,pace,shooting,passing,dribbling,defending,physic,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,attacking_volleys,skill_dribbling,skill_curve,skill_fk_accuracy,skill_long_passing,skill_ball_control,movement_acceleration,movement_sprint_speed,movement_agility,movement_reactions,movement_balance,power_shot_power,power_jumping,power_stamina,power_strength,power_long_shots,mentality_aggression,mentality_interceptions,mentality_positioning,mentality_vision,mentality_penalties,mentality_composure,defending_marking_awareness,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,goalkeeping_speed,ls,st,rs,lw,lf,cf,rf,rw,lam,cam,ram,lm,lcm,cm,rcm,rm,lwb,ldm,cdm,rdm,rwb,lb,lcb,cb,rcb,rb,gk,player_face_url,club_logo_url,club_flag_url,nation_logo_url,nation_flag_url
0,158023,https://sofifa.com/player/158023/lionel-messi/...,L. Messi,Lionel Andrés Messi Cuccittini,"RW, ST, CF",93,93,78000000.0,320000.0,34,1987-06-24,170,72,73.0,Paris Saint-Germain,French Ligue 1,1.0,RW,30.0,NaN,2021-08-10,2023.0,52,Argentina,1369.0,RW,10.0,Left,4,4,5,Medium/Low,Unique,Yes,144300000.0,"#Dribbler, #Distance Shooter, #FK Specialist, ...","Finesse Shot, Long Shot Taker (AI), Playmaker ...",85.0,92.0,91.0,95.0,34.0,65.0,85,95,70,91,88,96,93,94,91,96,91,80,91,94,95,86,68,72,69,94,44,40,93,95,75,96,20,35,24,6,11,15,14,8,NaN,89+3,89+3,89+3,92,93,93,93,92,93,93,93,91+2,87+3,87+3,87+3,91+2,66+3,64+3,64+3,64+3,66+3,61+3,50+3,50+3,50+3,61+3,19+3,https://cdn.sofifa.net/players/158/023/22_120.png,https://cdn.sofifa.net/teams/73/60.png,https://cdn.sofifa.net/flags/fr.png,https://cdn.sofifa.net/teams/1369/60.png,https://cdn.sofifa.net/flags/ar.png
1,188545,https://sofifa.com/player/188545/robert-lewand...,R. Lewandowski,Robert Lewandowski,ST,92,92,119500000.0,270000.0,32,1988-08-21,185,81,21.0,FC Bayern München,German 1. Bundesliga,1.0,ST,9.0,NaN,2014-07-01,2023.0,37,Poland,1353.0,RS,9.0,Right,4,4,5,High/Medium,Unique,Yes,197200000.0,"#Aerial Threat, #Distance Shooter, #Clinical F...","Solid Player, Finesse Shot, Outside Foot Shot,...",78.0,92.0,79.0,86.0,44.0,82.0,71,95,90,85,89,85,79,85,70,88,77,79,77,93,82,90,85,76,86,87,81,49,95,81,90,88,35,42,19,15,6,12,8,10,NaN,90+2,90+2,90+2,85,88,88,88,85,86+3,86+3,86+3,84+3,80+3,80+3,80+3,84+3,64+3,66+3,66+3,66+3,64+3,61+3,60+3,60+3,60+3,61+3,19+3,https://cdn.sofifa.net/players/188/545/22_120.png,https://cdn.sofifa.net/teams/21/60.png,https://cdn.sofifa.net/flags/de.png,https://cdn.sofifa.net/teams/1353/60.png,https://cdn.sofifa.net/flags/pl.png
2,20801,https://sofifa.com/player/20801/c-ronaldo-dos-...,Cristiano Ronaldo,Cristiano Ronaldo dos Santos Aveiro,"ST, LW",91,91,45000000.0,270000.0,36,1985-02-05,187,83,11.0,Manchester United,English Premier League,1.0,ST,7.0,NaN,2021-08-27,2023.0,38,Portugal,1354.0,ST,7.0,Right,4,5,5,High/Low,Unique,Yes,83300000.0,"#Aerial Threat, #Dribbler, #Distance Shooter, ...","Power Free-Kick, Flair, Long Shot Taker (AI), ...",87.0,94.0,80.0,88.0,34.0,75.0,87,95,90,80,86,88,81,84,77,88,85,88,86,94,74,94,95,77,77,93,63,29,95,76,88,95,24,32,24,7,11,15,14,11,NaN,90+1,90+1,90+1,88,89,89,89,88,86+3,86+3,86+3,86+3,78+3,78+3,78+3,86+3,63+3,59+3,59+3,59+3,63+3,60+3,53+3,53+3,53+3,60+3,20+3,https://cdn.sofifa.net/players/020/801/22_120.png,https://cdn.sofifa.net/teams/11/60.png,https://cdn.sofifa.net/flags/gb-eng.png,https://cdn.sofifa.net/teams/1354/60.png,https://cdn.sofifa.net/flags/pt.png


In [86]:
df.tail()

,sofifa_id,player_url,short_name,long_name,player_positions,overall,potential,value_eur,wage_eur,age,dob,height_cm,weight_kg,club_team_id,club_name,league_name,league_level,club_position,club_jersey_number,club_loaned_from,club_joined,club_contract_valid_until,nationality_id,nationality_name,nation_team_id,nation_position,nation_jersey_number,preferred_foot,weak_foot,skill_moves,international_reputation,work_rate,body_type,real_face,release_clause_eur,player_tags,player_traits,pace,shooting,passing,dribbling,defending,physic,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,attacking_volleys,skill_dribbling,skill_curve,skill_fk_accuracy,skill_long_passing,skill_ball_control,movement_acceleration,movement_sprint_speed,movement_agility,movement_reactions,movement_balance,power_shot_power,power_jumping,power_stamina,power_strength,power_long_shots,mentality_aggression,mentality_interceptions,mentality_positioning,mentality_vision,mentality_penalties,mentality_composure,defending_marking_awareness,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,goalkeeping_speed,ls,st,rs,lw,lf,cf,rf,rw,lam,cam,ram,lm,lcm,cm,rcm,rm,lwb,ldm,cdm,rdm,rwb,lb,lcb,cb,rcb,rb,gk,player_face_url,club_logo_url,club_flag_url,nation_logo_url,nation_flag_url
19234,261962,https://sofifa.com/player/261962/defu-song/220002,Song Defu,宋德福,CDM,47,52,70000.0,1000.0,22,1999-01-02,180,64,112541.0,Wuhan FC,Chinese Super League,1.0,SUB,36.0,NaN,2021-04-01,2021.0,155,China PR,NaN,NaN,NaN,Right,3,2,1,Medium/Medium,Lean (170-185),No,114000.0,NaN,NaN,58.0,35.0,46.0,48.0,42.0,49.0,46,32,48,50,30,45,33,38,48,49,56,60,55,53,70,46,62,51,46,30,52,42,38,43,42,37,38,43,48,6,10,5,15,13,NaN,44+2,44+2,44+2,46,45,45,45,46,46+2,46+2,46+2,47+2,46+2,46+2,46+2,47+2,48+2,46+2,46+2,46+2,48+2,48+2,46+2,46+2,46+2,48+2,15+2,https://cdn.sofifa.net/players/261/962/22_120.png,https://cdn.sofifa.net/teams/112541/60.png,https://cdn.sofifa.net/flags/cn.png,NaN,https://cdn.sofifa.net/flags/cn.png
19235,262040,https://sofifa.com/player/262040/caoimhin-port...,C. Porter,Caoimhin Porter,CM,47,59,110000.0,500.0,19,2002-03-18,175,70,445.0,Derry City,Rep. Ireland Airtricity League,1.0,RES,27.0,NaN,2021-02-22,2021.0,25,Republic of Ireland,NaN,NaN,NaN,Right,3,2,1,Medium/Medium,Lean (170-185),No,193000.0,NaN,NaN,59.0,39.0,50.0,46.0,41.0,51.0,54,33,46,51,32,41,53,31,50,42,60,58,64,49,69,49,47,63,47,38,49,39,51,49,44,47,37,44,47,11,12,6,8,10,NaN,45+2,45+2,45+2,47,46,46,46,47,47+2,47+2,47+2,49+2,47+2,47+2,47+2,49+2,49+2,46+2,46+2,46+2,49+2,48+2,44+2,44+2,44+2,48+2,14+2,https://cdn.sofifa.net/players/262/040/22_120.png,https://cdn.sofifa.net/teams/445/60.png,https://cdn.sofifa.net/flags/ie.png,NaN,https://cdn.sofifa.net/flags/ie.png
19236,262760,https://sofifa.com/player/262760/nathan-logue/...,N. Logue,Nathan Logue-Cunningham,CM,47,55,100000.0,500.0,21,2000-02-07,178,72,111131.0,Finn Harps,Rep. Ireland Airtricity League,1.0,RES,31.0,NaN,2021-06-18,2021.0,25,Republic of Ireland,NaN,NaN,NaN,Right,3,2,1,Medium/Medium,Lean (170-185),No,175000.0,NaN,NaN,60.0,37.0,45.0,49.0,41.0,52.0,39,32,43,49,37,47,37,37,49,49,60,60,58,46,59,50,57,56,50,34,51,38,45,46,39,36,38,44,48,8,6,7,10,6,NaN,45+2,45+2,45+2,46,46,46,46,46,47+2,47+2,47+2,48+2,47+2,47+2,47+2,48+2,47+2,46+2,46+2,46+2,47+2,47+2,45+2,45+2,45+2,47+2,12+2,https://cdn.sofifa.net/players/262/760/22_120.png,https://cdn.sofifa.net/teams/111131/60.png,https://cdn.sofifa.net/flags/ie.png,NaN,https://cdn.sofifa.net/flags/ie.png
19237,262820,https://sofifa.com/player/262820/luke-rudden/2...,L. Rudden,Luke Rudden,ST,47,60,110000.0,500.0,19,2002-04-23,173,66,111131.0,Finn Harps,Rep. Ireland Airtricity League,1.0,RES,12.0,NaN,2021-06-18,2021.0,25,Republic of Ireland,NaN,NaN,NaN,Right,3,2,1,Medium/Medium,Lean (170-185),No,239000.0,NaN,NaN,68.0,46.0,36.0,48.0,15.0,42.0,29,49,40,38,34,42,36,34,33,45,69,67,72,48,73,48,50,50,40,41,34,14,47

In [87]:
df.columns.tolist() 
#tolist() is used to show all the columns

['sofifa_id',
 'player_url',
 'short_name',
 'long_name',
 'player_positions',
 'overall',
 'potential',
 'value_eur',
 'wage_eur',
 'age',
 'dob',
 'height_cm',
 'weight_kg',
 'club_team_id',
 'club_name',
 'league_name',
 'league_level',
 'club_position',
 'club_jersey_number',
 'club_loaned_from',
 'club_joined',
 'club_contract_valid_until',
 'nationality_id',
 'nationality_name',
 'nation_team_id',
 'nation_position',
 'nation_jersey_number',
 'preferred_foot',
 'weak_foot',
 'skill_moves',
 'international_reputation',
 'work_rate',
 'body_type',
 'real_face',
 'release_clause_eur',
 'player_tags',
 'player_traits',
 'pace',
 'shooting',
 'passing',
 'dribbling',
 'defending',
 'physic',
 'attacking_crossing',
 'attacking_finishing',
 'attacking_heading_accuracy',
 'attacking_short_passing',
 'attacking_volleys',
 'skill_dribbling',
 'skill_curve',
 'skill_fk_accuracy',
 'skill_long_passing',
 'skill_ball_control',
 'movement_acceleration',
 'movement_sprint_speed',
 'movement_agi

In [88]:
df.describe()

,sofifa_id,overall,potential,value_eur,wage_eur,age,height_cm,weight_kg,club_team_id,league_level,club_jersey_number,club_contract_valid_until,nationality_id,nation_team_id,nation_jersey_number,weak_foot,skill_moves,international_reputation,release_clause_eur,pace,shooting,passing,dribbling,defending,physic,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,attacking_volleys,skill_dribbling,skill_curve,skill_fk_accuracy,skill_long_passing,skill_ball_control,movement_acceleration,movement_sprint_speed,movement_agility,movement_reactions,movement_balance,power_shot_power,power_jumping,power_stamina,power_strength,power_long_shots,mentality_aggression,mentality_interceptions,mentality_positioning,mentality_vision,mentality_penalties,mentality_composure,defending_marking_awareness,defending_standing_tackle,defending_sliding_tackle,goalkeeping_diving,goalkeeping_handling,goalkeeping_kicking,goalkeeping_positioning,goalkeeping_reflexes,goalkeeping_speed
count,19239.000000,19239.000000,19239.000000,1.916500e+04,19178.000000,19239.000000,19239.000000,19239.000000,19178.000000,19178.000000,19178.000000,19178.000000,19239.000000,759.000000,759.000000,19239.000000,19239.000000,19239.000000,1.806300e+04,17107.000000,17107.000000,17107.000000,17107.000000,17107.000000,17107.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,19239.000000,2132.000000
mean,231468.086959,65.772182,71.079370,2.850452e+06,9017.989363,25.210822,181.299704,74.943032,50580.498123,1.354364,20.945250,2022.764000,58.602682,14480.848485,12.567852,2.946151,2.352461,1.094184,5.374044e+06,68.213071,52.345297,57.312562,62.561174,51.703630,64.823289,49.577421,45.894433,51.783877,58.867977,42.463849,55.660429,47.268933,42.249025,53.072249,58.472010,64.652893,64.714902,63.500078,61.450023,64.068611,57.776860,64.813504,63.084880,65.007745,46.642705,55.538957,46.613545,50.330215,53.964603,47.858724,57.929830,46.601746,48.045584,45.906700,16.406102,16.192474,16.055356,16.229274,16.491814,36.439962
std,27039.717497,6.880232,6.086213,7.613700e+06,19470.176724,4.748235,6.863179,7.069434,54401.868535,0.747865,17.909369,1.213203,50.298614,35328.730217,7.039116,0.671560,0.767659,0.371098,1.494837e+07,10.933155,14.051623,10.068965,9.651312,16.189746,9.791886,18.034661,19.721023,17.294183,14.490858,17.653329,18.784590,18.181085,17.178590,15.026569,16.663722,15.167399,14.965426,14.862285,9.042281,14.324789,13.192224,12.122977,16.145279,12.663518,19.411583,16.972181,20.677077,19.621601,13.650481,15.768583,12.159326,20.200807,21.232718,20.755683,17.574028,16.839528,16.564554,17.059779,17.884833,10.751563
min,41.000000,47.000000,49.000000,9.000000e+03,500.000000,16.000000,155.000000,49.000000,1.000000,1.000000,1.000000,2021.000000,1.000000,1318.000000,1.000000,1.000000,1.000000,1.000000,1.600000e+04,28.000000,18.000000,25.000000,27.000000,14.000000,29.000000,6.000000,2.000000,5.000000,7.000000,3.000000,4.000000,6.000000,4.000000,9.000000,8.000000,14.000000,15.000000,18.000000,25.000000,15.000000,20.000000,22.000000,12.000000,19.000000,4.000000,10.000000,3.000000,2.000000,10.000000,7.000000,12.000000,4.000000,5.000000,5.000000,2.000000,2.000000,2.000000,2.000000,2.000000,15.000000
25%,214413.500000,61.000000,67.000000,4.750000e+05,1000.000000,21.000000,176.000000,70.000000,479.000000,1.000000,9.000000,2022.000000,21.000000,1338.000000,7.000000,3.000000,2.000000,1.000000,8.060000e+05,62.000000,42.000000,51.000000,57.000000,37.000000,59.000000,38.000000,30.000000,44.000000,54.000000,30.000000,50.000000,35.000000,31.000000,44.000000,55.000000,57.000000,58.000000,55.000000,56.000

In [89]:
df.dtypes

sofifa_id            int64
player_url          object
short_name          object
long_name           object
player_positions    object
                     ...  
player_face_url     object
club_logo_url       object
club_flag_url       object
nation_logo_url     object
nation_flag_url     object
Length: 110, dtype: object

In [90]:
df.size

2116290

In [91]:
df.dtypes.to_dict()

{'sofifa_id': dtype('int64'),
 'player_url': dtype('O'),
 'short_name': dtype('O'),
 'long_name': dtype('O'),
 'player_positions': dtype('O'),
 'overall': dtype('int64'),
 'potential': dtype('int64'),
 'value_eur': dtype('float64'),
 'wage_eur': dtype('float64'),
 'age': dtype('int64'),
 'dob': dtype('O'),
 'height_cm': dtype('int64'),
 'weight_kg': dtype('int64'),
 'club_team_id': dtype('float64'),
 'club_name': dtype('O'),
 'league_name': dtype('O'),
 'league_level': dtype('float64'),
 'club_position': dtype('O'),
 'club_jersey_number': dtype('float64'),
 'club_loaned_from': dtype('O'),
 'club_joined': dtype('O'),
 'club_contract_valid_until': dtype('float64'),
 'nationality_id': dtype('int64'),
 'nationality_name': dtype('O'),
 'nation_team_id': dtype('float64'),
 'nation_position': dtype('O'),
 'nation_jersey_number': dtype('float64'),
 'preferred_foot': dtype('O'),
 'weak_foot': dtype('int64'),
 'skill_moves': dtype('int64'),
 'international_reputation': dtype('int64'),
 'work_rat

In [92]:
df.isna().sum().to_dict()

{'sofifa_id': 0,
 'player_url': 0,
 'short_name': 0,
 'long_name': 0,
 'player_positions': 0,
 'overall': 0,
 'potential': 0,
 'value_eur': 74,
 'wage_eur': 61,
 'age': 0,
 'dob': 0,
 'height_cm': 0,
 'weight_kg': 0,
 'club_team_id': 61,
 'club_name': 61,
 'league_name': 61,
 'league_level': 61,
 'club_position': 61,
 'club_jersey_number': 61,
 'club_loaned_from': 18137,
 'club_joined': 1163,
 'club_contract_valid_until': 61,
 'nationality_id': 0,
 'nationality_name': 0,
 'nation_team_id': 18480,
 'nation_position': 18480,
 'nation_jersey_number': 18480,
 'preferred_foot': 0,
 'weak_foot': 0,
 'skill_moves': 0,
 'international_reputation': 0,
 'work_rate': 0,
 'body_type': 0,
 'real_face': 0,
 'release_clause_eur': 1176,
 'player_tags': 17798,
 'player_traits': 9841,
 'pace': 2132,
 'shooting': 2132,
 'passing': 2132,
 'dribbling': 2132,
 'defending': 2132,
 'physic': 2132,
 'attacking_crossing': 0,
 'attacking_finishing': 0,
 'attacking_heading_accuracy': 0,
 'attacking_short_passing'

## Dataset Preperation
* Dropping irrevelant columns and rows
* identifying duplicate columns
* Renaming Columns
* Feature Creation


### Removing irrevelant columns

In [93]:
df.columns.to_list()

['sofifa_id',
 'player_url',
 'short_name',
 'long_name',
 'player_positions',
 'overall',
 'potential',
 'value_eur',
 'wage_eur',
 'age',
 'dob',
 'height_cm',
 'weight_kg',
 'club_team_id',
 'club_name',
 'league_name',
 'league_level',
 'club_position',
 'club_jersey_number',
 'club_loaned_from',
 'club_joined',
 'club_contract_valid_until',
 'nationality_id',
 'nationality_name',
 'nation_team_id',
 'nation_position',
 'nation_jersey_number',
 'preferred_foot',
 'weak_foot',
 'skill_moves',
 'international_reputation',
 'work_rate',
 'body_type',
 'real_face',
 'release_clause_eur',
 'player_tags',
 'player_traits',
 'pace',
 'shooting',
 'passing',
 'dribbling',
 'defending',
 'physic',
 'attacking_crossing',
 'attacking_finishing',
 'attacking_heading_accuracy',
 'attacking_short_passing',
 'attacking_volleys',
 'skill_dribbling',
 'skill_curve',
 'skill_fk_accuracy',
 'skill_long_passing',
 'skill_ball_control',
 'movement_acceleration',
 'movement_sprint_speed',
 'movement_agi

In [94]:
df = df[['short_name',
 'player_positions',
 'value_eur','wage_eur',
 'age','height_cm','weight_kg',
 'club_name','league_name','club_loaned_from',
 'nationality_name',
 'preferred_foot',
 'pace','shooting','passing','dribbling','defending','physic','overall']].copy()

In [95]:
df.shape

(19239, 19)

In [96]:
df.isna().sum()

short_name              0
player_positions        0
value_eur              74
wage_eur               61
age                     0
height_cm               0
weight_kg               0
club_name              61
league_name            61
club_loaned_from    18137
nationality_name        0
preferred_foot          0
pace                 2132
shooting             2132
passing              2132
dribbling            2132
defending            2132
physic               2132
overall                 0
dtype: int64

### checking for duplicated rows 

In [97]:
df.loc[df.duplicated()]

,short_name,player_positions,value_eur,wage_eur,age,height_cm,weight_kg,club_name,league_name,club_loaned_from,nationality_name,preferred_foot,pace,shooting,passing,dribbling,defending,physic,overall


In [98]:
df.head()

,short_name,player_positions,value_eur,wage_eur,age,height_cm,weight_kg,club_name,league_name,club_loaned_from,nationality_name,preferred_foot,pace,shooting,passing,dribbling,defending,physic,overall
0,L. Messi,"RW, ST, CF",78000000.0,320000.0,34,170,72,Paris Saint-Germain,French Ligue 1,NaN,Argentina,Left,85.0,92.0,91.0,95.0,34.0,65.0,93
1,R. Lewandowski,ST,119500000.0,270000.0,32,185,81,FC Bayern München,German 1. Bundesliga,NaN,Poland,Right,78.0,92.0,79.0,86.0,44.0,82.0,92
2,Cristiano Ronaldo,"ST, LW",45000000.0,270000.0,36,187,83,Manchester United,English Premier League,NaN,Portugal,Right,87.0,94.0,80.0,88.0,34.0,75.0,91
3,Neymar Jr,"LW, CAM",129000000.0,270000.0,29,175,68,Paris Saint-Germain,French Ligue 1,NaN,Brazil,Right,91.0,83.0,86.0,94.0,37.0,63.0,91
4,K. De Bruyne,"CM, CAM",125500000.0,350000.0,30,181,70,Manchester City,English Premier League,NaN,Belgium,Right,76.0,86.0,93.0,88.0,64.0,78.0,91


#### There may be players with different id with same player. Filtering those by checking duplicacy of multiple attributes

In [99]:
df.loc[df.duplicated(subset=["short_name","overall","player_positions","age"])]

,short_name,player_positions,value_eur,wage_eur,age,height_cm,weight_kg,club_name,league_name,club_loaned_from,nationality_name,preferred_foot,pace,shooting,passing,dribbling,defending,physic,overall


### Renaming columns with appropriate format

In [100]:
df.columns

Index(['short_name', 'player_positions', 'value_eur', 'wage_eur', 'age',
       'height_cm', 'weight_kg', 'club_name', 'league_name',
       'club_loaned_from', 'nationality_name', 'preferred_foot', 'pace',
       'shooting', 'passing', 'dribbling', 'defending', 'physic', 'overall'],
      dtype='object')

In [101]:
df = df.rename(columns = {"short_name":"Short_Name",
                          "player_positions":"Position",
                          "value_eur":"Value_Eur",
                          "wage_eur":"Wage_Eur",
                          "age":"Age",
                          "height_cm":"Height_cm",
                          "weight_kg":"Weight_kg",
                          "club_name":"Club_Name",
                          "league_name":"League_Name",
                          "nationality_name":"Nationality",
                          "preferred_foot":"Preferred_Foot",
                          "pace":"Pace",
                          "shooting":"Shooting",
                          "passing":"Passing",
                          "dribbling":"Dribbling",
                          "defending":"Defending",
                          "physic":"Physic",
                          "overall":"Overall"
                         })

In [102]:
df.columns

Index(['Short_Name', 'Position', 'Value_Eur', 'Wage_Eur', 'Age', 'Height_cm',
       'Weight_kg', 'Club_Name', 'League_Name', 'club_loaned_from',
       'Nationality', 'Preferred_Foot', 'Pace', 'Shooting', 'Passing',
       'Dribbling', 'Defending', 'Physic', 'Overall'],
      dtype='object')

### Creating new feature is_loaned from the help of column club_loaned_from.

In [103]:
df.head()

,Short_Name,Position,Value_Eur,Wage_Eur,Age,Height_cm,Weight_kg,Club_Name,League_Name,club_loaned_from,Nationality,Preferred_Foot,Pace,Shooting,Passing,Dribbling,Defending,Physic,Overall
0,L. Messi,"RW, ST, CF",78000000.0,320000.0,34,170,72,Paris Saint-Germain,French Ligue 1,NaN,Argentina,Left,85.0,92.0,91.0,95.0,34.0,65.0,93
1,R. Lewandowski,ST,119500000.0,270000.0,32,185,81,FC Bayern München,German 1. Bundesliga,NaN,Poland,Right,78.0,92.0,79.0,86.0,44.0,82.0,92
2,Cristiano Ronaldo,"ST, LW",45000000.0,270000.0,36,187,83,Manchester United,English Premier League,NaN,Portugal,Right,87.0,94.0,80.0,88.0,34.0,75.0,91
3,Neymar Jr,"LW, CAM",129000000.0,270000.0,29,175,68,Paris Saint-Germain,French Ligue 1,NaN,Brazil,Right,91.0,83.0,86.0,94.0,37.0,63.0,91
4,K. De Bruyne,"CM, CAM",125500000.0,350000.0,30,181,70,Manchester City,English Premier League,NaN,Belgium,Right,76.0,86.0,93.0,88.0,64.0,78.0,91


In [104]:
df["Is_Loaned"] = df["club_loaned_from"].map(lambda x:0 if pd.isna(x) else 1)

#### Dropping the old column

In [105]:
df.drop(columns=["club_loaned_from"],inplace=True)

### creating new feeature No_Position representing number of position any player could play

In [106]:
df["No_Position"] = df["Position"].str.split(",").str.len()

In [107]:
df["No_Position"].head()

0    3
1    1
2    2
3    2
4    2
Name: No_Position, dtype: int64

### Handling missing values

In [109]:
df.isna().sum()

Short_Name           0
Position             0
Value_Eur           74
Wage_Eur            61
Age                  0
Height_cm            0
Weight_kg            0
Club_Name           61
League_Name         61
Nationality          0
Preferred_Foot       0
Pace              2132
Shooting          2132
Passing           2132
Dribbling         2132
Defending         2132
Physic            2132
Overall              0
Is_Loaned            0
No_Position          0
dtype: int64

#### There is common 61 rows with missing values of multiple features. Can be dropped as well but filling those missing values is possible.

In [73]:
df.loc[df["Wage_Eur"].isna()]

,Short_Name,Position,Value_Eur,Wage_Eur,Age,Height_cm,Weight_kg,Club_Name,League_Name,Nationality,Preferred_Foot,Pace,Shooting,Passing,Dribbling,Defending,Physic,Overall,Is_Loaned,No_Position
147,T. Alderweireld,CB,NaN,NaN,32,186,81,NaN,NaN,Belgium,Right,58.0,55.0,70.0,67.0,86.0,77.0,83,0,1
416,G. Krychowiak,"CDM, CM",NaN,NaN,31,187,84,NaN,NaN,Poland,Right,68.0,76.0,72.0,72.0,75.0,81.0,80,0,2
694,A. Dzyuba,ST,NaN,NaN,32,196,91,NaN,NaN,Russia,Right,65.0,78.0,69.0,71.0,29.0,86.0,78,0,1
926,M. Ozdoev,"CDM, CM",NaN,NaN,28,184,78,NaN,NaN,Russia,Right,75.0,70.0,72.0,76.0,73.0,76.0,77,0,2
1146,Y. Zhirkov,"LB, LM",NaN,NaN,37,180,75,NaN,NaN,Russia,Left,73.0,70.0,75.0,77.0,73.0,68.0,76,0,2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
12269,B. Sævarsson,RB,NaN,NaN,36,187,74,NaN,NaN,Iceland,Right,58.0,46.0,55.0,56.0,62.0,66.0,63,0,1
12891,O. Valakari,"CM, CAM",NaN,NaN,21,187,82,NaN,NaN,Finland,Left,62.0,62.0,62.0,58.0,47.0,68.0,63,0,2
14199,M. Stamenic,"CM, CDM, CAM",NaN,NaN,19,188,78,NaN,NaN,New Zealand,Right,68.0,51.0,62.0,68.0,52.0,65.0,62,0,3
14466,S. Roux,RB,NaN,NaN,28,181,71,NaN,NaN,New Zealand,Right,74.0,47.0,50.0,60.0,54.0,72.0,61,0,1


In [110]:
df["Club_Name"] = df["Club_Name"].fillna("Free Agent")
df["League_Name"] = df["League_Name"].fillna("No League")
df["Wage_Eur"] = df["Wage_Eur"].fillna(0)

In [113]:
df.loc[(df["Wage_Eur"]==0) & (df["Value_Eur"].isna() == True),"Value_Eur"] = 0df["Club_Name"] = df["Club_Name"].fillna("Free Agent")
df["League_Name"] = df["League_Name"].

In [114]:
df.isna().sum()

Short_Name           0
Position             0
Value_Eur           13
Wage_Eur             0
Age                  0
Height_cm            0
Weight_kg            0
Club_Name            0
League_Name          0
Nationality          0
Preferred_Foot       0
Pace              2132
Shooting          2132
Passing           2132
Dribbling         2132
Defending         2132
Physic            2132
Overall              0
Is_Loaned            0
No_Position          0
dtype: int64

In [117]:
df.loc[df["Value_Eur"].isna()]

,Short_Name,Position,Value_Eur,Wage_Eur,Age,Height_cm,Weight_kg,Club_Name,League_Name,Nationality,Preferred_Foot,Pace,Shooting,Passing,Dribbling,Defending,Physic,Overall,Is_Loaned,No_Position
1465,A. D'Alessandro,"CAM, RM",NaN,900.0,40,174,69,Club Nacional de Football,Uruguayan Primera División,Argentina,Left,42.0,71.0,80.0,75.0,38.0,50.0,75,0,2
1471,J. Sand,ST,NaN,13000.0,40,182,79,Club Atlético Lanús,Argentina Primera División,Argentina,Right,51.0,79.0,62.0,68.0,37.0,70.0,75,0,1
2429,M. Rodríguez,"LW, LM, ST",NaN,10000.0,40,173,74,Newell's Old Boys,Argentina Primera División,Argentina,Right,64.0,74.0,76.0,75.0,44.0,64.0,73,0,3
5469,Rubén Castro,ST,NaN,3000.0,40,169,68,FC Cartagena,Spanish Segunda División,Spain,Right,70.0,70.0,61.0,72.0,29.0,61.0,69,0,1
5479,L. Licht,"LB, LWB, LM",NaN,5000.0,40,174,72,Gimnasia y Esgrima La Plata,Argentina Primera División,Argentina,Left,64.0,61.0,65.0,72.0,68.0,66.0,69,0,3
7475,İ. Öztürk,CB,NaN,3000.0,40,186,80,Altay SK,Turkish Süper Lig,Turkey,Right,32.0,43.0,48.0,48.0,68.0,74.0,67,0,1
7478,P. Da Silva,CB,NaN,500.0,41,180,76,12 de Octubre FC,Paraguayan Primera División,Paraguay,Right,30.0,35.0,46.0,50.0,67.0,71.0,67,0,1
7488,Zheng Zhi,"CDM, CM",NaN,5000.0,40,180,75,Guangzhou FC,Chinese Super League,China PR,Right,35.0,64.0,67.0,66.0,67.0,59.0,67,0,2
12245,S. Nakamura,"CAM, CF, RW",NaN,800.0,43,178,71,Yokohama FC,Japanese J. League Division 1,Japan,Left,31.0,67.0,74.0,62.0,36.0,48.0,63,0,3
16209,K. Miura,ST,NaN,700.0,54,177,72,Yokohama FC,Japanese J. League Division 1,Japan,Right,30.0,63.0,55.0,62.0,19.0,46.0,59,0,1


In [120]:
missing_percentage = ((df.loc[df["Value_Eur"].isna()].shape[0])/(df.shape[0]))*100
missing_percentage

0.06757107957794065

#### only 0.067% of the total data are missing for the feature Value_Eur so dropping the missing values rows

In [121]:
df.dropna?

Signature:
df.dropna(
    *,
    axis: 'Axis' = 0,
    how: 'AnyAll | lib.NoDefault' = <no_default>,
    thresh: 'int | lib.NoDefault' = <no_default>,
    subset: 'IndexLabel | None' = None,
    inplace: 'bool' = False,
    ignore_index: 'bool' = False,
) -> 'DataFrame | None'
Docstring:
Remove missing values.

See the :ref:`User Guide <missing_data>` for more on which values are
considered missing, and how to work with missing data.

Parameters
----------
axis : {0 or 'index', 1 or 'columns'}, default 0
    Determine if rows or columns which contain missing values are
    removed.

    * 0, or 'index' : Drop rows which contain missing values.
    * 1, or 'columns' : Drop columns which contain missing value.

    Only a single axis is allowed.

how : {'any', 'all'}, default 'any'
    Determine if row or column is removed from DataFrame, when we have
    at least one NA or all NA.

    * 'any' : If any NA values are present, drop that row or column.
    * 'all' : If all values are NA, d

In [125]:
df.dropna(subset=["Value_Eur"], axis=0,inplace=True)

In [126]:
df.isna().sum()

Short_Name           0
Position             0
Value_Eur            0
Wage_Eur             0
Age                  0
Height_cm            0
Weight_kg            0
Club_Name            0
League_Name          0
Nationality          0
Preferred_Foot       0
Pace              2132
Shooting          2132
Passing           2132
Dribbling         2132
Defending         2132
Physic            2132
Overall              0
Is_Loaned            0
No_Position          0
dtype: int64

### There is still missing values of 2132 rows for multiple features. Figuring out those features.

In [132]:
missing = ["Pace","Physic","Shooting","Passing","Dribbling","Defending"]

In [138]:
df[missing].isna().sum()

Pace         2132
Physic       2132
Shooting     2132
Passing      2132
Dribbling    2132
Defending    2132
dtype: int64

In [134]:
for m in missing:
    print(df.loc[(df[m].isna()) & (df["Position"]=="GK")].shape)

(2132, 20)
(2132, 20)
(2132, 20)
(2132, 20)
(2132, 20)
(2132, 20)


In [139]:
df.loc[df["Position"]=="GK"].shape

(2132, 20)

### Figured out that all the 2132 missing values are of position GK

#### Splitting the dataset into two. For GK and for Players other than GK. Splitting is carried out inorder to analyze the GK analysis seperately than other players analysis

In [140]:
df_gk = df[df["Position"].str.contains("GK")].copy()
df_outfield = df[~df["Position"].str.contains("GK")].copy()

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


### Extracting Value and Wage Ration from Value_Eur and Wage_Eur

In [141]:
df["Value_Wage_Ratio"] = (df["Value_Eur"])/(df["Wage_Eur"])
df_outfield["Value_Wage_Ratio"] = (df_outfield["Value_Eur"])/(df_outfield["Wage_Eur"])

### datatype validation

In [142]:
df.dtypes

Short_Name           object
Position             object
Value_Eur           float64
Wage_Eur            float64
Age                   int64
Height_cm             int64
Weight_kg             int64
Club_Name            object
League_Name          object
Nationality          object
Preferred_Foot       object
Pace                float64
Shooting            float64
Passing             float64
Dribbling           float64
Defending           float64
Physic              float64
Overall               int64
Is_Loaned             int64
No_Position           int64
Value_Wage_Ratio    float64
dtype: object

# EDA

### For Target Feature

## What is the distribution of the overall rating is like?

In [ ]:
# for outfield
fig,axes = plt.subplots(nrows=1,ncols=2,figsize=(10,6))

df_outfield["Overall"].plot(kind="hist",title="Outfield Player's Overall Rating Distribution",ax=axes[0])
axes[0].set_xlabel("Outfield Rating")

df_gk["Overall"].plot(kind="hist",title="GK Player's Overall Rating Distribution",ax=axes[1])
axes[1].set_xlabel("GK Rating")

plt.tight_layout()
plt.show()


### => Both Outfield and GK Overall rating distributions are right skewed,
### meaning most players are average rated (60-65) with very few elite players.

## Are there any outliers in Overall Rating?

In [ ]:
# for outfield
fig , axes = plt.subplots(nrows=1,ncols=2 , figsize=(10,6))

df_outfield["Overall"].plot(kind="box",title="Outfield Rating",ax=axes[0])
axes[0].set_ylabel("Outfield Rating")

df_gk["Overall"].plot(kind="box",title="GK Rating",ax=axes[1])
axes[1].set_ylabel("GK Rating")
plt.tight_layout()
plt.show()

### => There is found to be both upper bound and lower bound outlier in outfield rating wheres there is only upper bound outlier in GK rating. Filtering out using IQR

In [ ]:
# ❌ IQR method was applied here but removed because Overall rating
# is a naturally skewed distribution where elite players like Messi, Ronaldo
# appear as statistical outliers but are valid and important data points.
# Outlier removal should only target data entry errors, not natural extremes.

In [ ]:
# # for outfield
# Q1_outfield = df_outfield["Overall"].quantile(0.25)
# Q3_outfield = df_outfield["Overall"].quantile(0.75)

# IQR_outfield = Q3_outfield-Q1_outfield

# upper_bound_outfield = Q3_outfield + (1.5*IQR_outfield)
# lower_bound_outfield = Q1_outfield - (1.5*IQR_outfield)

# df_outfield = df_outfield[(df_outfield["Overall"]<=upper_bound_outfield) & (df_outfield["Overall"]>=lower_bound_outfield)]

# # for GK
# Q1_gk = df_gk["Overall"].quantile(0.25)
# Q3_gk = df_gk["Overall"].quantile(0.75)

# IQR_gk = Q3_gk - Q1_gk

# upper_bound_gk = Q3_gk + (1.5 * IQR_gk)
# lower_bound_gk = Q1_gk - (1.5 * IQR_gk)

# df_gk = df_gk[(df_gk["Overall"]<=upper_bound_gk) & (df_gk["Overall"]>=lower_bound_gk)]

### For Numeric Features

In [ ]:
df_outfield.dtypes

In [ ]:
physical_feature = ["Age","Height_cm","Weight_kg"]


## **Physical Features**
### **AGE**
#### 1.What is the Age distribution of the Players?
#### 2. What is the common Age of the Players?
#### 3. Is there any Abnormal Age of any players?
### **HEIGHT**
#### 1. What is the Height distribution of players?
#### 2. Are there any abnormal Height of any players?
### **WEIGHT**
#### 1. What is the Weight Distribution of players?
#### 2. Are there any abnormal Weight of any players?

In [ ]:
fig,axes = plt.subplots(nrows=1,ncols=3,figsize=(12,6))
for i,col in enumerate(physical_feature):
    df_outfield[col].plot(kind="hist",ax=axes[i],title=f"Distribution of {col}")
    axes[i].set_xlabel(f"{col} of Outfield Players")
plt.tight_layout()
plt.show()

In [ ]:
df_outfield[(df_outfield["Age"]>45) | (df_outfield["Age"]<=15)]

In [ ]:
df_outfield["Age"].mode()[0]

In [ ]:
print(f"Mean : {df_outfield["Height_cm"].mean()}")
print(f"Median : {df_outfield["Height_cm"].median()}")

In [ ]:
df_outfield[df_outfield["Height_cm"]>200]

In [ ]:
df_outfield[(df_outfield["Weight_kg"]>100) | (df_outfield["Weight_kg"]<50)]

## **Physical Features**
### **AGE**
#### 1.What is the Age distribution of the Players?
#### => Its Right skewed which defines most of the players age ranges from 20-25 where there are some ages more than 50 as well making the distribution right skewed.
#### 2. What is the common Age of the Players?
#### => Common age of the players is 21.
#### 3. Is there any Abnormal Age of any players?
#### => Yes, There is abnormal age of one player with an age 54 who is currently playing in japanese leauge who belongs to japan.
### **HEIGHT**
#### 1. What is the Height distribution of players?
#### => Distribution of height is found to be symmetric or normal distribution. The mean and median of Height is also found to be almost same.
#### 2. Are there any abnormal Height of any players?
#### => There is no any abnormal height of any players though there are 7 players with more than 200cm whose position is ST and CB which totally makes sense for some  players.

### **WEIGHT**
#### 1. What is the Weight Distribution of players?
#### => The Distribution of symmetrical. It is bell curved
#### 2. Are there any abnormal Weight of any players?
#### => No there is no any abnormal weight. But there are 3 players more than 100kg but their height is above 170cm where being tall may have more wight makes it logical.  
#### and there is also 1 player less than 50kg but his height is also below 162cm which makes sense. 1 data can be consider as abnormal who has 110kg weight with only 178cm height

### Financial Features

In [ ]:
financial_feature = ["Value_Eur","Wage_Eur"]

## **Financial Features**
### **Value**
#### 1. What is the distrivution of Value_Eur?
#### 2. On what range does the most of the Value of the players lies?
#### 3. Are there any abnormal Market Value of any player?
### **Wage**
#### 1. What is the distribution of Wage_Eur?
#### 2. On what range does the most of the Wage of the Players lies?
#### 3. Are there any abnormal Wage of any players?

In [ ]:
fig,axes = plt.subplots(nrows=1,ncols=2,figsize=(12,6))
for i,col in enumerate(financial_feature):
    df_outfield[col].plot(kind="hist",ax=axes[i],title=f"Distribution of {col}")
    axes[i].set_xlabel(f"{col} of Outfield Players")
plt.tight_layout()
plt.show()

### For Market Value

In [ ]:
df_outfield["Log_Value_Eur"] = np.log1p(df_outfield["Value_Eur"])
df_outfield["Log_Value_Eur"].plot(kind="hist")

In [ ]:
top_expensive_players = df_outfield["Value_Eur"].sort_values(ascending=False).head()
top_expensive_players

In [ ]:
cheap_players = df_outfield["Value_Eur"].sort_values().head()
cheap_players

In [ ]:
Q1_value = df_outfield["Value_Eur"].quantile(0.25)
Q3_value = df_outfield["Value_Eur"].quantile(0.75)
print(f"From {Q1_value} to {Q3_value}")

### For Wage

In [ ]:
df_outfield["Log_Wage_Eur"] = np.log1p(df_outfield["Wage_Eur"])
df_outfield["Log_Wage_Eur"].plot(kind="hist")

In [ ]:
top_wage_players = df_outfield["Wage_Eur"].sort_values(ascending=False).head()
top_wage_players

In [ ]:
cheap_wage_players = df_outfield["Wage_Eur"].sort_values().head()
cheap_wage_players

In [ ]:
df_outfield["Wage_Eur"][df_outfield["Wage_Eur"]==500].count()

In [ ]:
Q1_wage = df_outfield["Wage_Eur"].quantile(0.25)
Q3_wage = df_outfield["Wage_Eur"].quantile(0.75)
print(f"From {Q1_wage} to {Q3_wage}")

## **Financial Features**
### **Value**
#### 1. What is the distribution of Value_Eur?
#### => Distribution is heavily right skewed (positive skew),indicating most players have low market value with very few elite players commanding extremely high values. ✅ Implemented log transformation to compress extreme values and make the distribution more symmetric for fair correlation analysis later
#### 2. On what range does the most of the Value of the players lies?
#### => most player's market value is between 525000 to 2100000
#### 3. Are there any abnormal Market Value of any player?
#### => there are no any abnormal market value. There is 2 low market value players i.e 15000 Eur which is less than the half of 3rd least market value player.but both are too old as well as they both reside from same leauge and nationality which might makes sense.
### **Wage**
#### 1. What is the distribution of Wage_Eur?
#### => Distribution is heavily right skewed(positive skew), indicating most of the players have low wages with very few elite players commanding extremely high wages. Implemented log transformation to compress extreme values and make the distribution more symmetric for fair correlation analysis later
#### 2. On what range does the most of the Wage of the Players lies?
#### => Most of the players have the range between 1000 to 9000 eur.
#### 3. Are there any abnormal Wage of any players?
#### => No there is no any abnormal wage. the lowest wage is 500 which is same for 1892 players which doesnt sound as an abnormal wage.

In [ ]:
fig, axes = plt.subplots(nrows=2, ncols=2, figsize=(12,8))

df_outfield["Value_Eur"].plot(kind="hist", ax=axes[0][0], title="Value_Eur (Original)")
axes[0][0].set_xlabel("Market Value (EUR)")

df_outfield["Log_Value_Eur"].plot(kind="hist", ax=axes[0][1], title="Value_Eur (Log Transformed)")
axes[0][1].set_xlabel("Log of Market Value")

df_outfield["Wage_Eur"].plot(kind="hist", ax=axes[1][0], title="Wage_Eur (Original)")
axes[1][0].set_xlabel("Wage (EUR)")

df_outfield["Log_Wage_Eur"].plot(kind="hist", ax=axes[1][1], title="Wage_Eur (Log Transformed)")
axes[1][1].set_xlabel("Log of Wage")

plt.tight_layout()
plt.show()

## **Skill Features**
### **Pace**
#### 1.What is the Pace distribution of the Players?
#### 2. What is the common Pace of the Players?
#### 3. Is there any Abnormal Pace of any players?
#### 4. Who has the highest Pace?
### **Shooting**
#### 1. What is the Shooting distribution of players?
#### 2. Are there any abnormal shooting attribute of any players?
### **Passing**
#### 1. What is the Passing Distribution of players?
#### 2. Are there any abnormal Passing attribute of any players?
#### 3. Who has the highest Passing attribute?
### **Dribbling**
#### 1. What is the Dribbling Distribution of players?
#### 2. Are there any abnormal Dribbling attribute of any players?
#### 3. Who has the highest Dribbling attribute?

### **Defending**
#### 1. What is the Defending Distribution of players?
#### 2. Are there any abnormal Defending attribute of any players?
#### 3. Who has the highest Defending attribute?
### **Physic**
#### 1. What is the Physic Distribution of players?
#### 2. Are there any abnormal Physic attribute of any players?
#### 3. Who has the highest Physic attribute?

In [ ]:
skill_feature = ["Pace","Shooting","Passing","Dribbling","Defending","Physic"]

In [ ]:
fig,axes = plt.subplots(nrows=2 ,ncols =3 ,figsize=(10,6))

df_outfield["Pace"].plot(kind="hist",title="Pace Distribution",ax=axes[0][0],bins=20)
axes[0][0].set_xlabel("Pace Attribute")

df_outfield["Shooting"].plot(kind="hist",title="Shooting Distribution",ax=axes[0][1],bins=20)
axes[0][1].set_xlabel("Shooting Attribute")

df_outfield["Passing"].plot(kind="hist",title="Passing Distribution",ax=axes[0][2],bins=20)
axes[0][2].set_xlabel("Passing Attribute")

df_outfield["Dribbling"].plot(kind="hist",title="Dribbling Distribution",ax=axes[1][0],bins=20)
axes[1][0].set_xlabel("Dribbling Attribute")

df_outfield["Defending"].plot(kind="hist",title="Defending Distribution",ax=axes[1][1],bins=20)
axes[1][1].set_xlabel("Defending Attribute")

df_outfield["Physic"].plot(kind="hist",title="Physic Distribution",ax=axes[1][2],bins=20)
axes[1][2].set_xlabel("Physic Attribute")

plt.tight_layout()
plt.show()

In [ ]:
#Common range of all attributes

Q1_pace = df_outfield["Pace"].quantile(0.25)
Q3_pace = df_outfield["Pace"].quantile(0.75)
print(f"Pace lies from {Q1_pace} to {Q3_pace}")

Q1_shooting = df_outfield["Shooting"].quantile(0.25)
Q3_shooting = df_outfield["Shooting"].quantile(0.75)
print(f"Shooting lies from {Q1_shooting} to {Q3_shooting}")

Q1_passing = df_outfield["Passing"].quantile(0.25)
Q3_passing = df_outfield["Passing"].quantile(0.75)
print(f"Passing lies from {Q1_passing} to {Q3_passing}")

Q1_dribbling = df_outfield["Dribbling"].quantile(0.25)
Q3_dribbling = df_outfield["Dribbling"].quantile(0.75)
print(f"dribbling lies from {Q1_dribbling} to {Q3_dribbling}")

Q1_defending = df_outfield["Defending"].quantile(0.25)
Q3_defending = df_outfield["Defending"].quantile(0.75)
print(f"defending lies from {Q1_defending} to {Q3_defending}")

Q1_physic = df_outfield["Physic"].quantile(0.25)
Q3_physic = df_outfield["Physic"].quantile(0.75)
print(f"Physic lies from {Q1_physic} to {Q3_physic}")

### Analyzing Pace attribute

In [ ]:
df_outfield[(df_outfield["Pace"]<30) | (df_outfield["Pace"]>=98)]

In [ ]:
striker = df_outfield["Position"].str.contains("ST",na=False,case=False)
df_outfield.groupby(striker)["Pace"].plot(kind="hist",alpha=0.5,title="ST VS Non-ST for Pace")
plt.legend(["Non-Striker","Striker"])

In [ ]:
highest_pace = df_outfield.loc[df_outfield["Pace"]==df_outfield["Pace"].max(),"Short_Name"]
highest_pace

In [ ]:
df.sort_values(by="Pace",ascending=False).head()

### Analyzing Shooting attribute

In [ ]:
center_back = df_outfield["Position"].str.contains("CB",case=False,na=False)
not_center_back = ~df_outfield["Position"].str.contains("CB",case=False,na=False)
df_outfield.groupby(center_back)["Shooting"].mean()

In [ ]:
df_outfield.groupby(center_back)["Shooting"].plot(kind="hist",alpha=0.5, title="CB VS Non-CB for Shooting")
plt.legend(["Non-CB","CB"])

In [ ]:
center_backs = df_outfield[center_back]
not_center_backs = df_outfield[not_center_back]

In [ ]:
df_outfield[(df_outfield["Shooting"]<25) | (df_outfield["Shooting"]>=98)]

In [ ]:
center_backs[center_backs["Shooting"]<25]

In [ ]:
not_center_backs[not_center_backs["Shooting"]<25]

In [ ]:
df_outfield.loc[df_outfield["Shooting"]==df_outfield["Shooting"].max(),["Short_Name","Shooting"]]

### Analyzing Passing attribute

In [ ]:
df_outfield[(df_outfield["Passing"]<=30) | (df_outfield["Passing"]>=95)]

In [ ]:
center_backs[center_backs["Passing"]<=30]

In [ ]:
not_center_backs[not_center_backs["Passing"]<=30]

In [ ]:
df_outfield.groupby(center_back)["Passing"].plot(kind="hist",alpha=0.5,title="CB VS Non-CB for Passing")
plt.legend(["Non-CB","CB"])

In [ ]:
df_outfield.loc[df_outfield["Passing"]==df_outfield["Passing"].max(),["Short_Name","Passing"]]

### Analysis on Dribbling Attribute

In [ ]:
df_outfield[df_outfield["Dribbling"]<29]

In [ ]:
df_outfield[df_outfield["Dribbling"]<35]

In [ ]:
center_backs[center_backs["Dribbling"]<35]

In [ ]:
not_center_backs[not_center_backs["Dribbling"]<35]

In [ ]:
df_outfield.groupby(center_back)["Dribbling"].plot(kind="hist",title="CB vs Non-CB for Dribbling",alpha=0.5)
plt.legend(["Non-CB","CB"])

In [ ]:
df_outfield.loc[df_outfield["Dribbling"]==df_outfield["Dribbling"].max(),["Short_Name","Dribbling"]]

### Analysis of Defending Attribute

In [ ]:
df_outfield[df_outfield["Defending"]<16]

In [ ]:
defender = df_outfield["Position"].str.contains("CB|LB|RB|LWB|RWB",na=False,case=False)
not_defender = ~df_outfield["Position"].str.contains("CB|LB|RB|LWB|RWB",na=False,case=False)
df_defender = df_outfield[defender]
df_not_defender = df_outfield[not_defender]
df_outfield.groupby(defender)["Defending"].mean()

In [ ]:
defender_with_cdm = df_outfield["Position"].str.contains("CB|LB|RB|LWB|RWB|CDM",na=False,case=False)
df_defender_with_cdm = df_outfield[defender_with_cdm]
df_outfield.groupby(defender_with_cdm)["Defending"].mean()

In [ ]:
df_defender_with_cdm

In [ ]:
df_outfield[df_outfield["Defending"]<16]

In [ ]:
df_outfield[df_outfield["Defending"]<25].shape

In [ ]:
df_not_defender[df_not_defender["Defending"]<25].shape

In [ ]:
df_defender[df_defender["Defending"]<25]

In [ ]:
df_outfield.groupby(defender_with_cdm)["Defending"].plot(kind="hist",alpha=0.5)
plt.legend(["Non-defender","Defender"])

In [ ]:
df_outfield[~defender_with_cdm].sort_values(by="Defending", ascending=False)[
    ["Short_Name", "Position", "Defending"]
].head(10)

In [ ]:
df_outfield.loc[df_outfield["Defending"]==df_outfield["Defending"].max(),["Short_Name","Position","Defending"]]

### Analysis of Physic Attribute

In [ ]:
df_outfield["Physic"].skew()

In [ ]:
Q1_physic = df_outfield["Physic"].quantile(0.25)
Q3_physic = df_outfield["Physic"].quantile(0.75)
print(f"{Q1_physic} to {Q3_physic}")

In [ ]:
df_outfield[df_outfield["Physic"]<32]

In [ ]:
df_outfield[df_outfield["Physic"]<34].shape

In [ ]:
df_outfield[(df_outfield["Age"]<21)&(df_outfield["Physic"]<34)].shape

In [ ]:
df_outfield.loc[df_outfield["Physic"]==df_outfield["Physic"].max(),["Short_Name","Physic","Position"]]

## **Skill Features**
### **Pace**
#### 1.What is the Pace distribution of the Players?
#### => most players have high pace (60-75) with very few having extremely low pace, creating a left skewed distribution
#### 2. Is there any Abnormal Pace of any players?
#### => There is not seem to be abnormal pace. The lowest pace attribute is 28 which is for 4 players. All of those 4 players have above 80kg weight and their age is also above 28 which makes the pace low.
#### 3. Who has the highest Pace?
#### => Kylian Mbappe with 97.

### **Shooting**
#### 1. What is the Shooting distribution of players?
#### => bimodal distribution with two peaks — one around 35-40 (defenders) and one around 58-62 (attackers/midfielders) which aligns with the CB vs non-CB average difference found
#### 2. Are there any abnormal shooting attribute of any players?
#### => There seems to be abnormal shooting attribute as 3 of the players have value 18. But they all are CB position so that may be somehow logical. Though the value is so low there are 287 values less than 25 and among 287 258 values are from players with cb and rest of the values are of full backs so they all are defenders.
#### 3. Who has the highest Shooting?
#### => Cristiano Ronaldo with 94.

### **Passing**
#### 1. What is the Passing Distribution of players?
#### => Passing is normally distributed with bell curved. This represent there is only few elite pass players and few weak pass players whereas most of the players resides between 51 to 64.
#### 2. Are there any abnormal Passing attribute of any players?
#### => There is no any abnormal passing attribute value of any players. The lowest passing value is 25 and all of them are again center backs. There is 84 players with less than 30 passing value and 81 of them are center backs whereas other 2 are full back and 1 of them is ST which logically sounds normal values.
#### 3. Who has the highest Passing attribute?
#### => De Bruyne with 93

### **Dribbling**
#### 1. What is the Dribbling Distribution of players?
#### => Dribbling attribute is slightly left skewed distribution indicating there are more players with weak dribbling than that of elite dribbling.
#### 2. Are there any abnormal Dribbling attribute of any players?
#### => There is not seem to be any abnormal value for dribbbling. The lowest value is 27 and that player is of 90kg with cb position so it sounds logical to be weak at dribbling. Similarly there are 112 players with less than 35 dribbling where 109 are CB and 3 of them are full backs.
#### 3. Who has the highest Dribbling attribute?
#### => Lionel Messi with 95. 

### **Defending**
#### 1. What is the Defending Distribution of players?
#### => Defending distribution is seem to be bimodal distribution with two parts - one around (20-35) for strikers or attackers and (55-65) for Defenders which aligns with the ST for non-ST average difference found. i.e Defenders mean is 61 whereas non-defenders mean is 51. filtering with defender with cdm mean is 62 and non-defender-withoutcdm is 50 which is slightly altered.
#### 2. Are there any abnormal Defending attribute of any players?
#### => The lowest value is 14 which sounds as abnormal but the there is only 1 player with 14 and 3 players with 15 where all of those players are ST position i.e attackers and also very young all resides between age of 17-19 where only one player is aged 22.  There are 1079 players with less than 25 whereas 1074 players are not defenders and 5 of those players are mostly wingers + full backs.
#### 3. Who has the highest Defending attribute?
#### => Van Dijk with 91.

### **Physic**
#### 1. What is the Physic Distribution of players?
#### => Bimodal distribution with two peaks around 20-35 (attackers) and 55-65 (defenders). Defenders average 62 while non-defenders average 50, confirming the two peaks represent distinct positional groups.
#### 2. Are there any abnormal Physic attribute of any players?
#### => The lowest physic value is 29 which seems to be normal as the player with 29 physic is 20 years old with 63 kg. There are 14 players with less than 34 and 10 of them are aged below 21 indicating most of the players with low physic are younger in age.
#### 3. Who has the highest Physic attribute?
#### => Casemiro with 90


# Bivariate Analysis

In [ ]:
df_outfield.columns

In [ ]:
# Extracting columns from position
df_outfield["Main_Position"] = df_outfield["Position"].str.split(",").str[0]


In [ ]:
df_outfield["Main_Position"].unique()

In [ ]:
position_groups = {
    "ST":"Attacker", "CF":"Attacker", "LW":"Attacker", "RW":"Attacker",
    "CAM":"Midfielder", "CM":"Midfielder", "CDM":"Midfielder",
    "LM":"Midfielder", "RM":"Midfielder",
    "CB":"Defender", "LB":"Defender", "RB":"Defender",
    "LWB":"Defender", "RWB":"Defender"
}
df_outfield["Grouped_Position"] =  df_outfield["Main_Position"].map(position_groups)

In [ ]:
df_outfield["Grouped_Position"].unique()

In [ ]:
numeric_columns=['Log_Value_Eur', 'Log_Wage_Eur' , 'Age', 'Height_cm',
       'Weight_kg', 'Pace', 'Shooting', 'Passing', 'Dribbling',
       'Defending', 'Physic']
categorical_columns = ['Preferred_Foot','Is_Loaned','No_Position',"Main_Position","Grouped_Position","League_Name","Nationality",]
target = "Overall"

In [ ]:
df_outfield["Log_Value_Eur"].head()

# Numeric Columns vs Overall Rating

In [ ]:
def numeric_vs_overall(df):
    for col in numeric_columns:
        plt.figure(figsize=(10,6))
        sns.scatterplot(data=df_outfield,x=col,y=target,alpha=0.4,hue="Grouped_Position")
        plt.title(f"{col} vs {target}")
        plt.tight_layout()
        plt.show()
numeric_vs_overall(df_outfield)

In [ ]:
hue_data = df_outfield["Main_Position"].apply(lambda x:"CB" if x=="CB" else "other")
plt.figure(figsize=(10,8))
sns.scatterplot(data=df_outfield,x="Height_cm",y="Overall",hue=hue_data)

In [ ]:
def Height_vs_Skills():
    for col in skill_feature:
        plt.figure(figsize=(10,6))
        sns.scatterplot(data=df_outfield,x="Height_cm",y=col,hue="Grouped_Position")
        plt.show()

Height_vs_Skills()

In [ ]:
skill_feature

In [ ]:

def weight_vs_skill():
    for col in skill_feature:
        plt.figure(figsize=(10,6))
        sns.scatterplot(data= df_outfield,x="Weight_kg",y=col,hue="Grouped_Position")
        plt.title(f"Weight vs {col}")
        plt.tight_layout()
        plt.show()

weight_vs_skill()

In [ ]:
df_outfield[numeric_columns + [target]].corr()["Overall"].sort_values(ascending=False)

In [ ]:
plt.figure(figsize=(12,10))
sns.heatmap(
    df_outfield[numeric_columns + [target]].corr(),
    annot=True,
    fmt='.2f',
    cmap='coolwarm',
    center=0
)
plt.title("Correlation Heatmap — Numeric Features vs Overall")
plt.tight_layout()
plt.show()

In [ ]:
# Log Value EUR and Log Wage EUR analysis

#Log_Value_Eur is showing the strong position relation with overall rating. As the correlation  is 0.90 which is highly correlated this might be the case of data leakage. 
#We are determining the features which will affect the overall rating but in this scenario Log_Value_Eur is affecting the overall rating positively but more than value affecting
#overall rating it seems to be overall rating which is affecting market value of any players which is data leakage in this scenario.
#Similar to Log_Wage_Eur, It shows 2 different pattern or trends. The Log_Wage_Eur above 9.5 is showing strong positive correlation whereas below 9.5 its showing no any relation.
#In below 9.5 the overall rating is not affected with wage value i.e Lower wage doesnt affects the players overall rating whereas higher wages are affecting the players overall raing.

#Both Log_Value_Eur and Log_Wage_Eur are likely results of Overall rating rather than causes of it.FIFA assigns market value and wage based on a players ability — meaning Overall 
#drives Value and Wage, not the other way around. Including these in a predictive model would cause data leakage and should be dropped during modelling phase.

#conclusion:
#These data leakage in modeling will create a problem where we are supposed to drop these columns in order to have greater accuracy.

In [ ]:
# Age analysis
#Age is seen as inverted U shaped with overall rating indicating younger and older players have less overall rating in compared to the age ranged between 22-35.

#conclusion:
#Young players and Old players have less overall rating compared to age ranged between 22-35. This feature is important for modeling.
                                                                                                                      
# Height analysis
#Height also shows the pattern like age with lower and higher height being less overall rating and middle ranged height being high as well as low rating. Most of the high 
#overall rating are of those players who are between middle ranged. 
#By further analyzing if the player is cb or not, figured out most of the tall players are from position cb but the overall rating being affecting is still the same

#Height doesnt affects the overall rating directly. But looking at height vs skill features there show negative relation to pace,passing,dribbling and positive relation to physic
#whereas those skill feature are directly dependent to overall rating. 
#This shows the indirect relationship between height and overall rating

#for example a player can be very tall so we cant directly say what will be his overall rating looking at it but we can say a play may have great physic with low dribbling,
#passing and pace which may define or predict the overall rating.

    
#conclusion:
#Height should be kept for modeling as it indirectly affects Overall through skill features like Physic, Pace and Dribbling rather than directly

#weight analysis
# Weight doesnt seem to have any direct relationship with overall rating as either weight is higher , lower or mid ranged there is the distribution of overall rating to low,high or mid ranged.
#But there is indirect relationship between overall rating and weight as weight is positively correlated with physic and negatively correlated with pace. whereas pace and physic 
#contain relationship with overall rating.

#conclusion:
#Wieght feature should be kept for modeling as it indirectly affects overall through features like physic and pace rather than directly
                                                                                                                      

In [ ]:
#bivariate analysis by position
df_defender = df_outfield[df_outfield["Grouped_Position"]=="Defender"]
df_attacker = df_outfield[df_outfield["Grouped_Position"]=="Attacker"]
df_midfielder = df_outfield[df_outfield["Grouped_Position"]=="Midfielder"]

## Defender Analysis

### for numerical feature

In [ ]:
#for defender
def defender_vs_numeric_features():
    for col in numeric_columns:
        plt.figure(figsize=(10,6))
        sns.scatterplot(data=df_defender,x=col,y="Overall",hue="Main_Position")
        plt.title(f"{col} vs Overall of Defenders")
        plt.tight_layout()
        plt.show()
defender_vs_numeric_features()

In [ ]:
df_defender[skill_feature + ["Overall"]].corr()["Overall"].sort_values(ascending=False)

In [ ]:
# Defender Analysis
#The positive relationship for defenders with overall is seen for physic,defending, passing, dribbling,  where fullbacks are found to be better on dribbilng,pace,shooting and
#cb are found to be better on defending, physic. 
#conclusion:
#1. value and wage must be dropped due to data leakage.
#2. age is inverted u shape i.e older and younger have low rating and mid ranged age has high rating. This feature should be kept
#3. weight and height doesnt have direct relationship but indirect relationship through skill features. These features should also be kept
#4. pace has weak relationship as its corr=0.16 vut it has weak only with cb but strong positive relationship with overall for full backs. This feature should also be kept
#5. all other skill feature such as shooting,passing,defending,dribbling,physic have positive correlation with overall rating
#6. Defending is one of the strongest feature among all with corr=0.91 and it is not leakage of data as well as defending is not determined by single overall feature
                                                            


### For Categorical Features

In [ ]:
def defender_vs_categorical():
    for col in categorical_columns:
        plt.figure(figsize=(10,6))
        sns.boxplot(data=df_defender,x=col,y="Overall")
        plt.title(f"Overall vs {col} for Defenders")
        plt.xticks(rotation=90)
        plt.tight_layout()
        plt.show()
defender_vs_categorical()

In [ ]:
categorical_columns

In [ ]:
top_leauges = ["English Premier League","French Ligue 1","Italian Serie A","German 1. Bundesliga","Spain Primera Division"]

In [ ]:
for col in top_leauges:
    print(f"{col} : {df_defender[df_defender["League_Name"]==col]["Overall"].median()}")
    print("")

defender analysis for categorical features
Players with right foot and left foot have almost same median,q1,q1,lower and upper whisker which shows right foot or left foot is directly not affecting the overall rating.
Though very few players i.e 2,3 players have higher as well as lower overall rating for right foot than left foot players it doesnt shows the relationship with overall rating.

Loaned player median, q1,q3 are slightly higher than unloaned players showing very less effect i.e loaned players being slighly higher rating on average. But most of the 
high rating players are from unloaned players concluding that the loan feature is not effecting the overall rating.

Defenders with 3 positions are slightly highly rated and are more concise than defender with 1 and 2 positions. Defenders with 1 and 2 position are wider meaning the rating is
distributed in wide range where 3 position is distributed in low range.

Main position and group position doesnt affects overall rating at all

Most of the premium defenders are found to be from the top 5 european leauges i.e premier leauge,bundesliga,laliga,serie A, Leauge 1. Premier leauge defenders are highly distributed
than that of other leauges. with their medians 

English Premier League : 75.0

French Ligue 1 : 72.0

Italian Serie A : 73.0

German 1. Bundesliga : 73.0

Spain Primera Division : 74.0

✅ "Top 5 European leagues have similar median 
    Overall (72-75) with Premier League marginally 
    highest at 75. The small difference suggests 
    league quality is relatively consistent across 
    top 5 European leagues rather than one league 
    being significantly stronger."



## Midfielder Analysis

In [ ]:
plt.figure(figsize=(10,6))
sns.scatterplot(data=df_midfielder,x=numeric_feature)